File: js/01-binary-tree/tests/binary-tree.test.js
const {
  createTree,
  degree,
  insert,
  remove,
  getFather,
  getBrother,
  getElement,
  calculateTreeDepth,
  calculateNodeLevel,
  toString
} = require('../src/binary-tree')

describe('binary-tree tests', () => {

  describe('createTree function', () => {

    test('create a new tree with an array of elements', () => {
      const elements = [6, 2, 8, 1, 4, 3]
      const tree = createTree(elements)
      const stringTree = toString(tree)

      expect(stringTree).toEqual("root:6 (left:2 (left:1 right:4 (left:3 ))right:8 )")
    })

    test('create a new tree with a single element', () => {
      const element = 6
      const tree = createTree(element)
      const stringTree = toString(tree)

      expect(stringTree).toEqual("root:6 ")
    })

  })

  describe('insert function', () => {

    test('perform a set of insertions', () => {
      const tree = createTree(6)

      let stringTree = toString(tree)
      expect(stringTree).toEqual("root:6 ")

      insert(tree, 2)
      stringTree = toString(tree)
      expect(stringTree).toEqual("root:6 (left:2 )")

      insert(tree, 8)
      stringTree = toString(tree)
      expect(stringTree).toEqual("root:6 (left:2 right:8 )")

      insert(tree, 1)
      stringTree = toString(tree)
      expect(stringTree).toEqual("root:6 (left:2 (left:1 )right:8 )")

      insert(tree, 1)
      stringTree = toString(tree)
      expect(stringTree).toEqual("root:6 (left:2 (left:1 )right:8 )")

      insert(tree, 4)
      stringTree = toString(tree)
      expect(stringTree).toEqual("root:6 (left:2 (left:1 right:4 )right:8 )")

      insert(tree, 3)
      stringTree = toString(tree)
      expect(stringTree).toEqual("root:6 (left:2 (left:1 right:4 (left:3 ))right:8 )")

      insert(tree, 3)
      stringTree = toString(tree)
      expect(stringTree).toEqual("root:6 (left:2 (left:1 right:4 (left:3 ))right:8 )")
    })

  })

  describe('remove function', () => {

    test('perform a set of removals', () => {
      const tree = createTree([37, 20, 10, 30, 80, 100, 5, 180, 90])

      let stringTree = toString(tree)
      expect(stringTree).toEqual("root:37 (left:20 (left:10 (left:5 )right:30 )right:80 (right:100 (left:90 right:180 )))")

      expect(remove(tree, 180)).toBeTruthy()
      stringTree = toString(tree)
      expect(stringTree).toEqual("root:37 (left:20 (left:10 (left:5 )right:30 )right:80 (right:100 (left:90 )))")

      remove(tree, 80)
      stringTree = toString(tree)
      expect(stringTree).toEqual("root:37 (left:20 (left:10 (left:5 )right:30 )right:100 (left:90 ))")

      remove(tree, 10)
      stringTree = toString(tree)
      expect(stringTree).toEqual("root:37 (left:20 (left:5 right:30 )right:100 (left:90 ))")

      remove(tree, 20)
      stringTree = toString(tree)
      expect(stringTree).toEqual("root:37 (left:30 (left:5 )right:100 (left:90 ))")

      remove(tree, 37)
      stringTree = toString(tree)
      expect(stringTree).toEqual("root:100 (left:90 (left:30 (left:5 )))")

      expect(remove(tree, 370)).toBeFalsy()
    })

  })

  describe('getElement function', () => {

    test('return an existing element', () => {
      const tree = createTree([6, 2, 8, 1, 4, 3])
      const node = getElement(tree, 8)

      expect(node.value).toEqual(8)
    })

    test('do not find an element', () => {
      const tree = createTree([6, 2, 8, 1, 4, 3])
      const node = getElement(tree, 10)

      expect(node).toBeUndefined()
    })

  })

  describe('degree function', () => {

    test('check nodes degrees', () => {
      const tree = createTree([6, 2, 8, 1, 4, 3])

      expect(degree(tree, 8)).toEqual(0)
      expect(degree(tree, 2)).toEqual(2)
      expect(degree(tree, 4)).toEqual(1)

      expect(degree(tree, 10)).toBeUndefined()
    })

  })

  describe('getFather function', () => {

    test('check fathers', () => {
      const tree = createTree([6, 2, 8, 1, 4, 3])

      expect(getFather(tree, 8).value).toEqual(6)
      expect(getFather(tree, 1).value).toEqual(2)

      expect(getFather(tree, 6)).toBeUndefined()
    })

  })

  describe('getBrother function', () => {

    test('check node brothers', () => {
      const tree = createTree([6, 2, 8, 1, 4, 3])

      expect(getBrother(tree, 8).value).toEqual(2)

      expect(getBrother(tree, 3)).toBeUndefined()
    })

  })

  describe('calculateTreeDepth function', () => {

    test('check tree depth', () => {
      const tree = createTree([6, 2, 8, 1, 4, 3])

      expect(calculateTreeDepth(tree)).toEqual(3)
    })

  })

  describe('calculateNodeLevel function', () => {

    test('check node levels', () => {
      const tree = createTree([6, 2, 8, 1, 4, 3])

      expect(calculateNodeLevel(tree, 8)).toEqual(1)
      expect(calculateNodeLevel(tree, 4)).toEqual(2)

      expect(calculateNodeLevel(tree, 40)).toBeUndefined()
    })

  })

})

File: java/binary-tree/src/test/java/br/com/gomide/binary_tree/BinaryTreeTest.java
package br.com.gomide.binary_tree;

import static org.junit.jupiter.api.Assertions.*;

import org.junit.jupiter.api.Test;

class BinaryTreeTest {

  @Test
  void createANewTree() {
    IBinaryTree<Integer> binaryTreeOps = new BinaryTree<>();
    Integer[] elements = new Integer[] { 6, 2, 8, 1, 4, 3 };

    Node<Integer> rootNode = binaryTreeOps.createTree(elements);

    assertEquals(binaryTreeOps.toString(rootNode), "root:6 (left:2 (left:1 right:4 (left:3 ))right:8 )");
  }

  @Test
  void performASetOfInsertions() {
    IBinaryTree<Integer> binaryTreeOps = new BinaryTree<>();
    Node<Integer> rootNode = binaryTreeOps.createTree(6);

    assertEquals(binaryTreeOps.toString(rootNode), "root:6 ");

    binaryTreeOps.insert(rootNode, 2);
    assertEquals(binaryTreeOps.toString(rootNode), "root:6 (left:2 )");

    binaryTreeOps.insert(rootNode, 8);
    assertEquals(binaryTreeOps.toString(rootNode), "root:6 (left:2 right:8 )");

    binaryTreeOps.insert(rootNode, 1);
    assertEquals(binaryTreeOps.toString(rootNode), "root:6 (left:2 (left:1 )right:8 )");

    binaryTreeOps.insert(rootNode, 1);
    assertEquals(binaryTreeOps.toString(rootNode), "root:6 (left:2 (left:1 )right:8 )");

    binaryTreeOps.insert(rootNode, 4);
    assertEquals(binaryTreeOps.toString(rootNode), "root:6 (left:2 (left:1 right:4 )right:8 )");

    binaryTreeOps.insert(rootNode, 3);
    assertEquals(binaryTreeOps.toString(rootNode), "root:6 (left:2 (left:1 right:4 (left:3 ))right:8 )");

    binaryTreeOps.insert(rootNode, 3);
    assertEquals(binaryTreeOps.toString(rootNode), "root:6 (left:2 (left:1 right:4 (left:3 ))right:8 )");
  }

  @Test
  void performASetOfRemovals() {
    IBinaryTree<Integer> binaryTreeOps = new BinaryTree<>();
    Integer[] elements = new Integer[] { 37, 20, 10, 30, 80, 100, 5, 180, 90 };

    Node<Integer> rootNode = binaryTreeOps.createTree(elements);

    assertEquals(binaryTreeOps.toString(rootNode),
        "root:37 (left:20 (left:10 (left:5 )right:30 )right:80 (right:100 (left:90 right:180 )))");

    binaryTreeOps.remove(rootNode, 180);
    assertEquals(binaryTreeOps.toString(rootNode),
        "root:37 (left:20 (left:10 (left:5 )right:30 )right:80 (right:100 (left:90 )))");

    binaryTreeOps.remove(rootNode, 80);
    assertEquals(binaryTreeOps.toString(rootNode),
        "root:37 (left:20 (left:10 (left:5 )right:30 )right:100 (left:90 ))");

    binaryTreeOps.remove(rootNode, 10);
    assertEquals(binaryTreeOps.toString(rootNode), "root:37 (left:20 (left:5 right:30 )right:100 (left:90 ))");

    binaryTreeOps.remove(rootNode, 20);
    assertEquals(binaryTreeOps.toString(rootNode), "root:37 (left:30 (left:5 )right:100 (left:90 ))");

    binaryTreeOps.remove(rootNode, 37);
    assertEquals(binaryTreeOps.toString(rootNode), "root:100 (left:90 (left:30 (left:5 )))");
  }

  @Test
  void getAnElement() {
    IBinaryTree<Integer> binaryTreeOps = new BinaryTree<>();
    Integer[] elements = new Integer[] { 6, 2, 8, 1, 4, 3 };

    Node<Integer> rootNode = binaryTreeOps.createTree(elements);

    Node<Integer> node = binaryTreeOps.getByElement(rootNode, 8);

    assertEquals(node.getValue(), 8);
  }

  @Test
  void dontFindRequiredElement() {
    IBinaryTree<Integer> binaryTreeOps = new BinaryTree<>();
    Integer[] elements = new Integer[] { 6, 2, 8, 1, 4, 3 };

    Node<Integer> rootNode = binaryTreeOps.createTree(elements);

    Node<Integer> node = binaryTreeOps.getByElement(rootNode, 10);

    assertNull(node);
  }

  @Test
  void checkNodeDegree() {
    IBinaryTree<Integer> binaryTreeOps = new BinaryTree<>();
    Integer[] elements = new Integer[] { 6, 2, 8, 1, 4, 3 };

    Node<Integer> rootNode = binaryTreeOps.createTree(elements);

    assertEquals(binaryTreeOps.degree(rootNode, 8), 0);
    assertEquals(binaryTreeOps.degree(rootNode, 2), 2);
    assertEquals(binaryTreeOps.degree(rootNode, 4), 1);
    assertNull(binaryTreeOps.degree(rootNode, 10));
  }

  @Test
  void getFather() {
    IBinaryTree<Integer> binaryTreeOps = new BinaryTree<>();
    Integer[] elements = new Integer[] { 6, 2, 8, 1, 4, 3 };

    Node<Integer> rootNode = binaryTreeOps.createTree(elements);

    assertEquals(binaryTreeOps.getFather(rootNode, 8).getValue(), 6);
    assertEquals(binaryTreeOps.getFather(rootNode, 1).getValue(), 2);
    assertNull(binaryTreeOps.getFather(rootNode, 6));
  }

  @Test
  void getBrother() {
    IBinaryTree<Integer> binaryTreeOps = new BinaryTree<>();
    Integer[] elements = new Integer[] { 6, 2, 8, 1, 4, 3 };

    Node<Integer> rootNode = binaryTreeOps.createTree(elements);

    assertEquals(binaryTreeOps.getBrother(rootNode, 8).getValue(), 2);
    assertNull(binaryTreeOps.getBrother(rootNode, 3));
  }

  @Test
  void calculateTreeDepths() {
    IBinaryTree<Integer> binaryTreeOps = new BinaryTree<>();
    Integer[] elements = new Integer[] { 6, 2, 8, 1, 4, 3 };

    Node<Integer> rootNode = binaryTreeOps.createTree(elements);

    assertEquals(binaryTreeOps.calculateTreeDepth(rootNode), 3);
  }

  @Test
  void calculateNodeLevel() {
    IBinaryTree<Integer> binaryTreeOps = new BinaryTree<>();
    Integer[] elements = new Integer[] { 6, 2, 8, 1, 4, 3 };

    Node<Integer> rootNode = binaryTreeOps.createTree(elements);

    assertEquals(binaryTreeOps.calculateNodeLevel(rootNode, 8), 1);
    assertEquals(binaryTreeOps.calculateNodeLevel(rootNode, 4), 2);
    assertNull(binaryTreeOps.calculateNodeLevel(rootNode, null));
  }

}
